### ISSUE TITLE
Suggestion: Auto-reset Locked Weapons on Level Exit

### ISSUE THREAD

**Issue by RPGHacker** [MegamixEngine/MegamixEngine#195]
MaGMML2 featured a bunch of levels in which some (or all) of Mega Man's special weapons were locked. Sometimes this was to prevent the player from breaking the level, other times it was a gameplay element and players could unlock the weapons again by finding certain items in the level. Something similar can be achieved in the Megamix Engine via the setWeaponLocked() function. For example, here I lock all weapons but the Mega Buster:

```
// Lock all weapons but the mega buster
for (var i = 0; i < array_length_1d(global.weaponObject); ++i)
{
    setWeaponLocked(global.weaponObject[i], true);
}

setWeaponLocked(objBusterShot, false);
```

However, upon doing this and exiting the level (be it via the Exit option or by Game Over), you will find that all of the weapons that were locked by the level remain locked afterwards, even after entering another level.

Suggestion: to prevent a user's level from breaking engine state, there should be some kind of auto-reset feature for locked weapons. For example: all weapons could automatically be reset to their state from the beginning of the level upon exiting. However, this might not be ideal since it would make it impossible to permanently unlock weapons inside levels. Probably a better solution would be to add a third, optional parameter to setWeaponLocked() called autoReset. When set to true, that respective weapon could automatically be reset upon exiting the level.

Note that it's likely possible to reset the state of locked weapons directly from a level's code, however, since there are multiple ways of leaving a level, I suspect that this is risky and that it's easy for some way to slip through and for weapons to remain locked after leaving a level. A feature directly in the engine should be safer and therefore preferable here.

**Comment by MegamixEngine** [comment]
You must lock weapons in the object "objWeaponSettings". This is how MaGMML2 reset them. Having weapons unlock between rooms would completely break how unlocking weapons in the Example Game works, etc..

**Comment by RPGHacker** [comment]
Thanks for the quick help. I swear I looked into weapon-related code for a lengthy amount of time and was not able to locate this object.

However, this doesn't seem to solve the problem. I now removed the code I previously had in the level and instead placed an objWeaponSettings in the level with the following instance creation code:

```
for (var i = 0; i <= global.totalWeapons; i += 1)
{
    global.weaponLocked[i] = true;
}

global.weaponLocked[global.weaponID[? objBusterShot]] = false;
```

When entering this level, my weapons get locked as they're supposed to. However, upon exiting the (level (in this case via Pause -> Exit), I find that the respective weapons still remain locked once entering another level, and I do think I know why. I think it's a bug in this object's room end code. In the create code, the comment tells you to set `global.weaponLocked` to lock specific weapons. This makes a lot of sense, to separate it from the `global.weaponUnlocked` array and thus make it possible to lock weapons only temporarily. However, in the room end code of that object, instead of resetting `global.weaponLocked` to false, it actually resets `global.weaponUnlocked` to true. This means that all weapons you locked via `global.weaponLocked` actually remain locked in the next level, and as another side-effect, all weapons that were previously (permanently) locked are now (permanently) unlocked.

I think the room end code in objWeaponSettings should actually be something like

```
for (i = 0; i <= global.totalWeapons; i += 1)
{
    global.weaponLocked[i] = false;
    global.infiniteEnergy[i] = false;
}
```
